### ISSUE TITLE
Backspace misbehavior in wrapped textarea

### ISSUE THREAD

**Issue by ExoticMatter** [mikke89/RmlUi#313]
When a textarea contains wrapped text, backspace causes the caret to move to the left, as if the left arrow key was pressed immediately after pressing backspace. Every prior wrapped line causes the caret to move left an additional extra character.

Example text (text is one single line; each string of characters is separated by a space to allow wrapping; `|` represents the caret)

    aaaaaaaaaaaaaaaaaaaa
    aaaaaaaaaaaaaaaaaaaa
    aaaaaaaaaaaaaaaaaaaa
    aaaaaaaaaaaaaaaaabcd|

After pressing backspace:

    ...aaaaaaa|abc

(redundant text omitted for brevity)

**Comment by mikke89** [comment]
Thanks for the well written report, I am able to replicate this on my end.

I found that this also occurs if we select some text on a soft-wrapped line and then move the caret left or right. I will look into it.

**Comment by ExoticMatter** [comment]
I was previously under the impression that the delete key didn't cause the caret to move left after a wrapped line, but it in fact does.

**Comment by mikke89** [comment]
Text editing is quite tricky it turns out. But I've made an effort to fix this. I had to refactor a bit, trying to simplify things, because it was very hard to follow the old code with a lot of duplicate state.

I may have accidently introduced new bugs with these changes, because [text editing is tricky](https://lord.io/text-editing-hates-you-too/). Could you test the changes I made in the `textarea` branch, see if it fixes the issue and try to break everything and discover any new (or old) bugs?

**Comment by ExoticMatter** [comment]
It has fixed the issue, and I haven't found any new problems with the patch so far (or old problems, for that matter.)

**Comment by mikke89** [comment]
I've been working more on the text widget. I discovered some issues introduced with the previous changes, mainly to do with password fields.

In any case, I've made a lot of improvements since then, both quality-of-life improvements, fixing small issues, and code cleanup. See the commit messages for details: https://github.com/mikke89/RmlUi/commits/textarea.

I'm re-opening this for some time so that people can be aware of these changes and report any related issues or other feedback.
